scripts/hca-utils/hca_utils/utils.py

```diff
@@ -0,0 +1,332 @@
+import csv
+from functools import lru_cache
+import os
+from requests.exceptions import HTTPError
+from typing import Set
+import urllib.parse
+
+import google.auth
+from google.auth.transport.requests import AuthorizedSession
+from google.cloud import bigquery, storage
+from requests_cache.core import CachedSession
+
+
+class HcaUtils:
+    def __init__(self, environment: str, project: str, dataset: str):
+        self.environment = environment
+
+        if environment == "dev":
+            self.project = "broad-jade-dev-data"
+        else:
+            self.project = project
+
+        self.dataset = dataset
+
+        self.filename_template = f"sd-{project}-{dataset}-{{table}}.csv"
+
+        self.bigquery_client = bigquery.Client(project=self.project)
+
+        bucket_projects = {"prod": "mystical-slate-284720",
+                           "dev": "broad-dsp-monster-hca-dev"}
+        self.bucket_project = bucket_projects[environment]
+        self.bucket = f"broad-dsp-monster-hca-{environment}-staging-storage"
+
+        jade_urls = {"prod": "https://jade-terra.datarepo-prod.broadinstitute.org",
+                     "dev": "https://jade.datarepo-dev.broadinstitute.org"}
+        self.base_url = jade_urls[environment]
+
+        # this is always prod for bigquery, storage, etc.
+        gcp_creds, _ = google.auth.load_credentials_from_file("path/to/prod/credentials.json")
+        self.gcp_creds = gcp_creds
+
+        # this depends on the Jade env to interact with
+        creds_path = {"prod": "path/to/prod/credentials.json",
+                      "dev": "path/to/dev/credentials.json"}
+        jade_creds, _ = google.auth.load_credentials_from_file(creds_path[environment])
+        self.jade_creds = jade_creds
+
+    # bigquery interactions
+    def get_all_table_names(self) -> Set[str]:
+        """
+        Gets the table names for the target dataset.
+        :return: A set of table names.
+        """
+        query = f"""
+        SELECT table_name
+        FROM `{self.project}.datarepo_{self.dataset}.INFORMATION_SCHEMA.TABLES` WHERE table_type = "VIEW"
+        """
+
+        return self._hit_bigquery(query)
+
+    def get_file_table_names(self) -> Set[str]:
+        """
+        Gets the table names for tables that have a `file_id` column.
+        :return: A set of table names.
+        """
+        query = f"""
+        WITH fileRefTables AS (SELECT * FROM `{self.project}.datarepo_{self.dataset}.INFORMATION_SCHEMA.COLUMNS` WHERE column_name = "file_id"),
+        desiredViews AS (SELECT * FROM `{self.project}.datarepo_{self.dataset}.INFORMATION_SCHEMA.TABLES` WHERE table_type = "VIEW")
+        SELECT desiredViews.table_name FROM fileRefTables JOIN desiredViews ON fileRefTables.table_name = desiredViews.table_name
+        """
+
+        return self._hit_bigquery(query)
+
+    def get_duplicates(self, target_table: str) -> Set[str]:
```

Comment:
The query in this function is A LOT so please ask questions; I tried to use clear names and commented with my thought process

Reply:
i think it might help readability/extensibility to pull the shared snippets out into helper functions, or even build out an entirely separate `BigqueryQueryBuilder` class. that way you'd be able to use descriptive function names to make these verbose comments less necessary.

Reply:
+1, I think it'd help focus things to extract the db concerns into a separate class.